File: 0636_exclusive-time-of-functions.cpp
class Solution {
private:
    const char MID = 'm';
    const char END = 'e';
public:
//     void printStack(stack<pair<int, pair<char, int>>> st, string s) {
//         while (!st.empty()) {
//             auto top = st.top();
//             st.pop();
//             printf("%s %d %c %d\n", s.c_str(), top.first, top.second.first, top.second.second);
//         }
//         puts("");
//     }
    
//     void printVector(vector<int>& vi) {
//         for (int i : vi) {
//             printf("%d ", i);
//         }
//         puts("");
//     }
    
    vector<int> exclusiveTime(int n, vector<string>& logs) {
        vector<int> vi(n, 0);
        stack<pair<int, pair<char, int>>> st;
        for (string log : logs) {
            // cout << log << endl;
            // printStack(st, "");
            size_t p1 = log.find(":");
            int id = stoi(log.substr(0, p1));
            
            size_t p2 = log.find(":", p1 + 1);
            char ch = log[p1 + 1];
            
            int time = stoi(log.substr(p2 + 1));
            
            if (ch == END) {
                int subtime = 0;
                while (st.top().second.first == MID) {
                    auto top = st.top();
                    st.pop();
                    subtime += top.second.second;
                }
                auto top = st.top();
                st.pop();
                int dur = time - subtime - top.second.second + 1;
                vi[id] += dur;
                st.push({id, {MID, dur + subtime}});
            } else {
                st.push({id, {ch, time}});
            }
            // printStack(st, "\t");
            // printVector(vi);
            // puts("---------------------------------");
        }
        return vi;
    }
};

File: 0636_exclusive-time-of-functions.py
class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        dct = [0] * n
        stack = collections.deque([])
        start_time = None
        for log in logs:
            tokens = log.split(':')
            func_id, timestamp = int(tokens[0]), int(tokens[2])
            if not stack:
                stack.append((func_id, tokens[1], timestamp))
            elif stack[-1][0] == func_id and stack[-1][1] != tokens[1]:
                timestamp += 1
                stack.pop()
                dct[func_id] += timestamp - start_time
            else:
                prev_func_id = stack[-1][0]
                dct[prev_func_id] += timestamp - start_time
                stack.append((func_id, tokens[1], timestamp))
            start_time = timestamp
        return dct
